README.md
# cookbook-zabbix-cookbook

This cookbook is supposed to install the Zabbix server and agent. A direct rewrite of [this cookbook](https://github.com/nollieheel/ga/tree/master/cookbooks/chef-zabbix).

For now, there is only the `agent_package` recipe, which installs the agent. Server installation is future TODO.

## Supported Platforms

Redhat, Centos, Ubuntu

## Attributes

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['listen_port']</tt></td>
    <td>String</td>
    <td>Port number where the Zabbix agent will listen on.</td>
    <td><tt>'10050'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['hostname']</tt></td>
    <td>String</td>
    <td>Human-friendly name for a monitored Zabbix node, also known in the Zabbix UI as hostname.</td>
    <td><tt>'Some Zabbix Node'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['server_ip']</tt></td>
    <td>String</td>
    <td>Location of the Zabbix server/s. Multiple IPs can be comma-separated.</td>
    <td><tt>'127.0.0.1'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['enable_active_checks']</tt></td>
    <td>Boolean</td>
    <td>Whether to enable active checks for this particular node or not.</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['refresh_active_checks']</tt></td>
    <td>Integer</td>
    <td>Frequency (in seconds) of active checks, if enabled.</td>
    <td><tt>120</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['enable_remote_commands']</tt></td>
    <td>String</td>
    <td>Whether to enable remote commands from the Zabbix server. Accepts the strings '1' for true, or '0' for false.</td>
    <td><tt>'0'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['advanced_params']</tt></td>
    <td>Hash</td>
    <td>Key-value pairs of strings to be included in the agent config file as advanced parameters.</td>
    <td><tt>{}</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['user_params']</tt></td>
    <td>Hash</td>
    <td>Key-value pairs of strings to be included in the agent config file as user parameters.</td>
    <td><tt>{}</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-zabbix']['agent']['packages']</tt></td>
    <td>Array</td>
    <td>Array of packages to be installed by an agent.</td>
    <td><tt>['zabbix-agent', 'zabbix-sender']</tt></td>
  </tr>
</table>

## Usage

### cookbook-zabbix::agent_package

After specifying the proper agent attributes, especially `server_ip` and `hostname`, include `cookbook-zabbix::agent_package` in your node's `run_list` to install the agent:

```json
{
  "run_list": [
    "recipe[cookbook-zabbix::agent_package]"
  ]
}
```

## License and Authors

Author:: Earth U (<iskitingbords @ gmail.com>)


CHANGELOG.md
# 1.0.1

Verified support for Ubuntu 12.04.

# 1.0.0

Install zabbix-sender on client nodes. Add additional repos for RHEL platforms.

# 0.1.0

Initial release of cookbook-zabbix. Just the agent recipe so far.


recipes/agent_config.rb
#
# Author:: Earth U (<iskitingbords @ gmail.com>)
# Cookbook Name:: cookbook-zabbix
# Recipe:: agent_config
#
# Copyright 2016, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

aagent = node['cookbook-zabbix']['agent']

template '/etc/zabbix/zabbix_agentd.conf' do
  only_if  { Dir.exist?('/etc/zabbix') }
  action   :create
  mode     0644
  notifies :restart, "service[#{aagent['service_name']}]"

  variables(
    :server_ip              => aagent['server_ip'],
    :listen_port            => aagent['listen_port'],
    :enable_active_checks   => aagent['enable_active_checks'],
    :hostname               => aagent['hostname'],
    :refresh_active_checks  => aagent['refresh_active_checks'],
    :enable_remote_commands => aagent['enable_remote_commands'],
    :advanced_params        => aagent['advanced_params'],
    :user_params            => aagent['user_params']
  )
end


recipes/install_repo.rb
#
# Author:: Earth U (<iskitingbords @ gmail.com>)
# Cookbook Name:: cookbook-zabbix
# Recipe:: install_repo
#
# Copyright 2016, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'apt::default' if node['platform_family'] == 'debian'

node['cookbook-zabbix']['repos'].each do |r|
  fpath = "#{Chef::Config[:file_cache_path]}/#{r[:repo].split('/').last}"

  remote_file fpath do
    source r[:repo]
    action :create_if_missing
  end

  case node['platform_family']
  when 'rhel'
    yum_package r[:name] do
      source fpath
      action :install
    end

  when 'debian'
    dpkg_package r[:name] do
      source   fpath
      action   :install
      notifies :run, 'execute[apt-get update]', :immediately
    end
  end
end


recipes/agent_package.rb
#
# Author:: Earth U (<iskitingbords @ gmail.com>)
# Cookbook Name:: cookbook-zabbix
# Recipe:: agent_package
#
# Copyright 2016, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# NOTE:
# Clients will have to manually open their port 10050 before
# using this recipe. It can't be included here because iptables
# setup in different systems vary too much.

include_recipe "#{cookbook_name}::install_repo"

package node['cookbook-zabbix']['agent']['packages'] do
  if node['platform_family'] == 'rhel'
    flush_cache(
      :before => true
    )
  end
end

service node['cookbook-zabbix']['agent']['service_name'] do
  if node['platform_family'] == 'rhel'
    restart_command '/sbin/service zabbix-agent try-restart'
  end
  supports [:start, :stop, :restart, :status]
  action   [:enable, :start]
end

include_recipe "#{cookbook_name}::agent_config"


attributes/default.rb
#
# Author:: Earth U (<iskitingbords @ gmail.com>)
# Cookbook Name:: cookbook-zabbix
# Attribute:: default
#
# Copyright 2016, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

cb = 'cookbook-zabbix'

# The repository url and filename will determine the version of Zabbix.
# Version 2.2 is only used on Ubuntu 12.04
repo_24 = 'http://repo.zabbix.com/zabbix/2.4'
repo_22 = 'http://repo.zabbix.com/zabbix/2.2'
main_repo = value_for_platform(
  ['redhat', 'centos', 'amazon'] => {
    'default' => {
      :repo => "#{repo_24}/rhel/6/x86_64/zabbix-release-2.4-1.el6.noarch.rpm",
      :name => 'zabbix-release-2.4-1.el6.noarch'
    }
  },
  'debian' => {
    'default' => {
      :repo => "#{repo_24}/debian/pool/main/z/zabbix-release/zabbix-release_2.4-1+wheezy_all.deb",
      :name => 'zabbix-release'
    }
  },
  'ubuntu' => {
    '12.04' => {
      :repo => "#{repo_22}/ubuntu/pool/main/z/zabbix-release/zabbix-release_2.2-1+precise_all.deb",
      :name => 'zabbix-release'
    },
    'default' => {
      :repo => "#{repo_24}/ubuntu/pool/main/z/zabbix-release/zabbix-release_2.4-1+trusty_all.deb",
      :name => 'zabbix-release'
    }
  }
)

default[cb]['repos'] = value_for_platform_family(
  'rhel' => [
    main_repo,
    {
      :repo => "#{repo_24}/rhel/6/x86_64/zabbix-get-2.4.1-1.el6.x86_64.rpm",
      :name => 'zabbix-get-2.4.1-1.el6.x86_64'
    },
    {
      :repo => "#{repo_24}/rhel/6/x86_64/zabbix-sender-2.4.1-1.el6.x86_64.rpm",
      :name => 'zabbix-sender-2.4.1-1.el6.x86_64'
    }
  ],
  'debian' => [main_repo]
)

default[cb]['server']['packages'] = [
  value_for_platform_family(
    'rhel'   => 'zabbix-web-mysql',
    'debian' => 'zabbix-frontend-php'
  ),
  'zabbix-server-mysql',
  'zabbix-get'
]

default[cb]['agent']['packages'] = [
  'zabbix-agent',
  'zabbix-sender'
]
default[cb]['agent']['service_name'] = 'zabbix-agent'

default[cb]['agent']['listen_port'] = '10050'
default[cb]['agent']['hostname']    = 'Some Zabbix Node'
default[cb]['agent']['server_ip']   = '127.0.0.1'

default[cb]['agent']['enable_active_checks']   = true
default[cb]['agent']['refresh_active_checks']  = 120
default[cb]['agent']['enable_remote_commands'] = '0'

default[cb]['agent']['advanced_params'] = {}
default[cb]['agent']['user_params']     = {}
